### ISSUE TITLE
FiroPoWSCC and Zpool issue continues

### ISSUE THREAD

**Issue by Boulder08** [RainbowMiner/RainbowMiner#2824]
Some changes were done in https://github.com/RainbowMiner/RainbowMiner/issues/2806 but still plenty of rejected shares. Often 30-50% failure rate and the watchdog steps in every now and then.

09:16:58.101 GPU[00-SCC]:   Rejected #20, RT in 68 ms. [Invalid share(25)][firopow.eu.mine.zpool.ca:1326]

**Comment by Ondalf** [comment]
Thing is...

8GB cards are now dying when allocating DAG for sccpow (the actual name for the scc, see the [README.md](https://github.com/stakecube/StakeCubeCoin)) with 1300 (default firo) DAG length.

Pool always provides correct seed_hash in every mining.notify that defines the 3240 length DAG for the hasher, but only few miners actually respects it. GMiner is another in addition to TTminer that has it working properly.
Pools backend has verifier done for both of these cases, since wallet does not really check the seedhash/mixhash, if they are right.
There are 2 different backends behind single port where you connect and its random where you end up at. 10 consecutive rejects and pool disconns usually forcing to use the another backend making this sccpow universally supported to any firopow miners. Be it 1300 or 3240 blocks DAG increment. Technically there was a point when wallet was accepting any DAGs (yes, even 000...).

What I could suggest instead of limiting the miner programs -- how about creating sccpow as algorithm within RBM, so we can finally rebrand the firopow from poolside? And like SCC documents states -- the name really is SCCpow, not FiroPowSCC. To expand this, it needs just list of supported miners to be launched with -a firopow unless there is naming done properly.

And regarding reject spree, since the backend you end up is random, i've had a good 15 minutes time wasted to rejects with single 1080 in hopes to get into correct one. Since limit is 10 rejects before disconn, it takes a while.

Instead, what this system allowed - ever since sccpow was introduced, we had ways to mine it regardless the miner had support or not. And wallet accepted them. Hash algorithm matches, DAG length not. But seems 8GB cards are now out of question with miners that doesn't really respect the supplied seedhash from mining.notify.

So what you did earlier axing all but TTminers off was probably not good way to handle it... But now you have description why rejectspree happens, and why its more or less feature for compatibility in mind instead of error.

TL;DR:
- this is pools feature to make __all firopow__ miners to hash sccpow
- but it needs time and luck to make yourself to end to correct endpoint, which you cannot alter
- get us properly named algo with curated list of miners and I'll do a port without this fallback feature

**Comment by Ondalf** [comment]
Added xnonce2 check to stratum -- end result: TTminer is broken. And now automatically rejected.

Only miners that respects the stratum protocol so far (xnonce2 compared to submitted nonce) are t-rex/24.8+, GMiner/3.44+, TRM/0.10.7+, TBMiner/2.25+, nanominer/3.7.7+.
No other submits valid shares after checking if the pool distributed xnonce2 is within the first left 4 hexas in submitted nonce.

So good try. Next time try to respect the stratum protocol as well. More about this (for miner writers) in [nicehash kawpow protocol definition](https://github.com/nicehash/Specifications/blob/master/KawPow_NiceHash_v1.0.txt) AND stratumv1 definitions.

**Comment by Ondalf** [comment]
So from testing this matter out and seeing how clients are connected, and which backend they use

Miners **minimum** tested version and backend they prefer at time writing:
prefers firopow mode (epoch 7xx):
t-rex/24.8
teamredminer/0.10.5.1
TBMiner/2.25
WildRig/0.40.5W

prefers sccpow mode (epoch 2xx):
GMiner/3.43
nanominer/3.7.7-cuda11

**Comment by Ondalf** [comment]
So if you mind, undo [this](https://github.com/RainbowMiner/RainbowMiner/commit/0ce50b5ce1879b29add1aa99976dbe2111dea448) as you simply allowed miner that submits invalid shares towards pool ignoring working miners.

**Comment by RainbowMiner** [comment]
Thank you very much for the infos. I will update the source.

**Comment by RainbowMiner** [comment]
Now I do get rejected shares all over - no matter which miner I am using (tested with Gminer, Nanominer and SccMiner - all calculate the DAG size correctly, but then produce rejected shares, only). Maybe it is time to remove firopow and use sccpow only on Zpool. I am now already using algorithm "SCCPow".

**Comment by Ondalf** [comment]
10 consecutive rejects and you get randomized another backend.
there is 2 stratums with different verifiers - one as workaround for miners that ignores seed_hash and uses height to calculate dag based on 1300 firo defined height. so any pure firo miners will work once you end up to correct one.
and another backend that is supplied with 3240 length dag with verifiers made for it.

granted - both backends supplies only 3240 seed_hash to steer miners that has it right to correct.
while height concerned miners are steered to the 1300 backend.

let it sit for a while so it throws you around. any of those miners mentioned earlier are guaranteed to at some point get to correct one, but since its always 50/50 chance on connect, which backend you get greeted at, it may take a bit.

**Comment by Ondalf** [comment]
Poolside user error fixed for non-0x-prefixed submits
SRB works for both firo and scc without any special modifications at zpool (updated the list as well for working miners) - miner reads pool sent seed_hash, so it can adapt using -a firopow to sccpow.

**Comment by UselessGuru** [comment]
> Maybe it is time to remove firopow and use sccpow only on Zpool. I am now already using algorithm "SCCPow".

@Ondalf 
I second this. Rejects are not the way to get the miner get switched to the corret endpoint.
Why not adding SSCpow as a separate algo and port which then points to the working endpoint?

**Comment by Ondalf** [comment]
> > Maybe it is time to remove firopow and use sccpow only on Zpool. I am now already using algorithm "SCCPow".
> 
> @Ondalf I second this. Rejects are not the way to get the miner get switched to the correct endpoint. Why not adding SSCpow as a separate algo with its own port which then points to the working endpoint?

Hmmm... I'll figure out something. But yes, algoname change has been in the work for ages now. Only thing I've seen over time is that SCC didn't really push the algo namechange, when it happened...

PS. this dual backend solution works perfectly for literally any firopow miners.

**Comment by UselessGuru** [comment]
@Ondalf 

Any news?

**Comment by Ondalf** [comment]
> @Ondalf
> 
> Any news?

Not yet. Got distracted with LCN with flex and webinterface improvements and forgot this matter again... Will try to solve it during this week.
But the idea I had was to simply create dedicated ports for sccpow and firopow and list SCC under both (so it will be just copypaste for api/status line, but internally it would be same exact thing.
This should address the issue of having correct backend regardless the port.

But I still don't see an issue with this as older miners still knows the algorithm name as "firopow" so it will be a bit weird to suggest people use "hey, we got this SCC coin here, and if you got gminer, use -a firopow and point it to sccpow". What this causes is a mess with customer support.

Regardless, i'll try to break them apart - haven't just yet figured clever way to do it so yiimp understands, that they're essentially same algos...

Again, people with more than 8GB vRAM has zero issues with whatever miner they use... 10 rejects and they get directed to working backend, as you **really cannot figure out the miner until it handshakes the version**.
I COULD make it reject on spot the miner, if it vanders to wrong backend automatically, but not too sure about it. Dropped the initial difficulty from 16 to 8, which makes the wandering time before kick shorter. With high qty of cards, the switch should happen pretty fast anyways.

**Comment by UselessGuru** [comment]
> Regardless, i'll try to break them apart - haven't just yet figured clever way to do it so yiimp understands, that they're essentially same algos...

Any news?

**Comment by UselessGuru** [comment]
@Ondalf 

Any news?

**Comment by Ondalf** [comment]
> @Ondalf
> 
> Any news?

Most likely will roll the renaming update tomorrow. Did the required stratum work earlier this week in preparation few other algos as well, that uses kawpow -alike chatter.

**Comment by Ondalf** [comment]
Changes applied. SCC shares same sharepool in both algos.
Remove that rewrite of firopow->sccpow for zpool immediately, so we get users with proper miners and less VRAM to the game ^^

**Comment by Ondalf** [comment]
You may want to add sccpow to this *1000 multiplier as well, since sccpow miner shows 17kh on 1080:
https://github.com/RainbowMiner/RainbowMiner/blob/master/Modules/MinerAPIs.psm1#L1136

**Comment by Boulder08** [comment]
@RainbowMiner Looks like there's something to be tuned with Firopow mining (this error is from benchmarking SrbMinerMulti.

[2024-09-08 12:05:32] GPU [BUS:09] not enough free memory [need 7694 MB : have 6044 MB] [0]
